### PR TITLE
Add bridge support (obfs4 or user-supplied)

### PR DESCRIPTION
**Pull request by mig5**
I took a stab at #435

It works, but I need some help with the widgets in the SettingsDialog. When choosing 'custom' bridges, the textbox doesn't expand the parent widgets, so is hard to read. (If you are able to paste in some contents and hit Save, then next time you open the SettingsDialog, it is much better because the settings were detected and so displayed by default).

Additionally it's possible to mess up the other Tor connection widgets above (control port, socket file) when they show the hidden extra options. I couldn't work out how to make them all place nicely :( but otherwise the feature should work.

I think it's the Buttons HBox that refuses to budge when child widgets in the VBoxes above it, show or hide hidden widgets....

I copied the obfs4 transports that seem to come hardcoded into Tor Browser.

Maybe you can fork what I've done and fix the widget issues if it's something trivial that I'm missing.

### CODE REVIEW

**Comment by mig5** [comment]
Think I've fixed the widget issues, by simple fact that the Bridges widgets should only be shown when 'bundled' Tor is selected anyway.

**Comment by micahflee** [comment]
This is excellent! So I've found 2 issues:

For some reason, I can't connect using obfs4. I get 10% through, and then it never gets past that. I'm currently on a crappy wifi network, but I tried tethering to my phone as well, and same issue.

Using custom bridges works great. However, there's no error checking. If you type something that's not in the proper format, like `foobar` in the custom bridge text field, it silently throws an exception:

```
user@dev ~/code/onionshare (mig5-435_support_bridges) $ ./dev_scripts/onionshare-gui --debug
Onionshare 1.1 | https://onionshare.org/
[Jan 13 2018 03:40:37 PM] Onion.__init__
[Jan 13 2018 03:40:37 PM] OnionShare.__init__
[Jan 13 2018 03:40:37 PM] OnionShareGui.__init__
[Jan 13 2018 03:40:37 PM] Settings.__init__
[Jan 13 2018 03:40:37 PM] Settings.load
[Jan 13 2018 03:40:37 PM] Settings.load: Trying to load /home/user/.config/onionshare/onionshare.json
[Jan 13 2018 03:40:37 PM] TorConnectionDialog.__init__
[Jan 13 2018 03:40:37 PM] TorConnectionDialog.start
[Jan 13 2018 03:40:37 PM] TorConnectionThread.__init__
[Jan 13 2018 03:40:37 PM] TorConnectionThread.run
[Jan 13 2018 03:40:37 PM] Onion.connect
[Jan 13 2018 03:40:37 PM] TorConnectionDialog._connected_to_tor
[Jan 13 2018 03:40:38 PM] OnionShareGui.open_settings
[Jan 13 2018 03:40:38 PM] SettingsDialog.__init__
[Jan 13 2018 03:40:38 PM] Settings.__init__
[Jan 13 2018 03:40:38 PM] Settings.load
[Jan 13 2018 03:40:38 PM] Settings.load: Trying to load /home/user/.config/onionshare/onionshare.json
[Jan 13 2018 03:40:38 PM] SettingsDialog._update_autoupdate_timestamp
[Jan 13 2018 03:40:38 PM] SettingsDialog.connection_type_automatic_toggled
[Jan 13 2018 03:40:38 PM] SettingsDialog.authenticate_no_auth_toggled
[Jan 13 2018 03:40:39 PM] SettingsDialog.connection_type_automatic_toggled
[Jan 13 2018 03:40:39 PM] SettingsDialog.connection_type_bundled_toggled
[Jan 13 2018 03:42:37 PM] SettingsDialog.test_tor_clicked
[Jan 13 2018 03:42:37 PM] SettingsDialog.settings_from_fields
[Jan 13 2018 03:42:37 PM] Settings.__init__
[Jan 13 2018 03:42:37 PM] Settings.load
[Jan 13 2018 03:42:37 PM] Settings.load: Trying to load /home/user/.config/onionshare/onionshare.json
[Jan 13 2018 03:42:37 PM] SettingsDialog._disable_buttons
[Jan 13 2018 03:42:37 PM] Onion.__init__
[Jan 13 2018 03:42:37 PM] Onion.connect
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/stem/socket.py", line 446, in _make_socket
    control_socket.connect(self._socket_path)
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/code/onionshare/onionshare/onion.py", line 237, in connect
    self.c = Controller.from_socket_file(path=self.tor_control_socket)
  File "/usr/lib/python3.6/site-packages/stem/control.py", line 1042, in from_socket_file
    control_socket = stem.socket.ControlSocketFile(path)
  File "/usr/lib/python3.6/site-packages/stem/socket.py", line 429, in __init__
    self.connect()
  File "/usr/lib/python3.6/site-packages/stem/socket.py", line 247, in connect
    self._socket = self._make_socket()
  File "/usr/lib/python3.6/site-packages/stem/socket.py", line 449, in _make_socket
    raise stem.SocketError(exc)
stem.SocketError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/user/code/onionshare/onionshare_gui/settings_dialog.py", line 510, in test_tor_clicked
    onion.connect(settings=settings, config=self.config, tor_status_update_func=tor_status_update_func)
  File "/home/user/code/onionshare/onionshare/onion.py", line 240, in connect
    raise BundledTorBroken(strings._('settings_error_bundled_tor_broken', True).format(e.args[0]))
onionshare.onion.BundledTorBroken: Something is wrong with OnionShare connecting to Tor in the background:
[Errno 2] No such file or directory
```

There some be some error checking to make sure it's in the right format, and throw an error if it's not.

Also, if I add custom bridges that are in the correct format, but that aren't actually Tor bridges (like `8.8.8.8:443 722EF1F0939E2F1863035CF72D862525CC5F9AA3`), connecting to Tor gets to 10% and then never proceeds.

You said you copied the obfs4 bridges hardcoded into Tor Browser. Where in the Tor Browser directory did you find those? And how often does Tor Browser update that hardcoded list? I wonder if maybe my problem is that it's trying to connect to bridges that don't exist anymore.

**Comment by micahflee** [comment]
Oh also this branch now has a simple merge conflict. I resolved it like this (increasing the timeouts from 45s and 60s to 120s and 150s) so feel free to make your own commit to resolve it like this if you want:

```
commit 92449085ef581808c2ae65599c85009efa6b6d43 (HEAD -> mig5-435_support_bridges)
Merge: 2dee9f8 890acf9
Author: Micah Lee <micah@micahflee.com>
Date:   Sat Jan 13 15:17:47 2018 -0800

    Merge branch '435_support_bridges' of https://github.com/mig5/onionshare into mig5-435_support_bridges
    
    Also, increase Tor connection timeouts.

diff --cc onionshare/onion.py
index 50cfa03,e76072c..ef11e74
--- a/onionshare/onion.py
+++ b/onionshare/onion.py
@@@ -254,8 -264,13 +264,13 @@@ class Onion(object)
                      break
                  time.sleep(0.2)
  
-                 # Timeout after 90 seconds
-                 if time.time() - start_ts > 90:
+                 # If using bridges, it might take a bit longer to connect to Tor
+                 if self.settings.get('tor_bridges_use_custom_bridges') or self.settings.get('tor_bridges_use_obfs4'):
 -                    connect_timeout = 60
++                    connect_timeout = 150
+                 else:
 -                    # Timeout after 45 seconds
 -                    connect_timeout = 45
++                    # Timeout after 120 seconds
++                    connect_timeout = 120
+                 if time.time() - start_ts > connect_timeout:
                      print("")
                      self.tor_proc.terminate()
                      raise BundledTorTimeout(strings._('settings_error_bundled_tor_timeout'))
```

**Comment by mig5** [comment]
Thanks for the review.

I got the obfs4 bridges in two ways:

1) First I configured obfs4 in my Tor Browser, and then obtained the resulting data from the torrc within Tor Browser's config (on my Mac, Library/Application Support/TorBrowser-Data/Tor/torrc

2) Then from https://gitweb.torproject.org/builders/tor-browser-bundle.git/tree/Bundle-Data/PTConfigs/bridge_prefs.js (which seems to show the same ones)

obfs4 worked for me though it took a long time from memory

Thanks for the review, I'll look at sanitisation of user-supplied bridges

**Comment by micahflee** [comment]
Just adding a comment here, too. The QLabel with a link in it should have the same fix as in #521.

**Comment by mig5** [comment]
@micahflee the timeout conflict is fixed, and I have added some basic regex checking of custom bridges. If a custom bridge line is invalid, it is simply ignored. If *no* custom bridges are valid, an alert is shown, and then the bridges setting is undone to just be back to normal bundled mode (so that having clicked 'Save', Tor can actually try to reconnect with default settings like it normally would).

Don't merge this through yet, though, for the following reasons:

1) I think obfs4 in torrc doesn't work out of the box.. deeper debugging of the bootstrap phase suggested that it was iterating through each bridge but 'ignoring' them, until the timeout runs out. Reading up on it, I think it's necessary to install the 'obfs4proxy' binary and then add the torrc setting ```ClientTransportPlugin obfs4 exec /usr/local/bin/obfs4proxy``` (for example). See the README at https://github.com/Yawning/obfs4.

My guess is that Tor Browser, since it doesn't add a ClientTransportPlugin line, has its own obfs4proxy somehow built in when using obfs4.. but I might be wrong about all this.


2) While testing on macOS, I seem to *still* see some funky UI widget bugs related to the position of the 'Test Tor Button' overlapping over the Custom Bridges textarea widget. I would appreciate some testing once you have access to your Mac, as those widgets are hard to negotiate. I'm pretty sure it was OK on Linux (or you'd have noticed already in your earlier test), must be something OS specific.. 

Perhaps we could do the next release with 'custom bridges' support and then add obfs4 later. If I'm right that obfs4 requires extra binaries installed on the machine, I'm not sure how we'd work out whether the user has adequately installed such dependencies or not yet, or whether you'd want to package onionshare with its own 'bundled' obfs4proxy.

**Comment by Baccount** [comment]
after I choose use custom bridges, the "test tor settings" button covers some of the textfield. Increasing the "settings GUI" size will fix the issue. shown at 5-10 seconds in the attached video.

When I click "Use obfs4 pluggable transports" and click "Don't use bridges", the gray box will incorrectly resize it self. shown at 1-2 seconds.
[bugs.mov.zip](https://github.com/micahflee/onionshare/files/1629203/bugs.mov.zip)

**Comment by mig5** [comment]
I believe I've fixed the overlap button issue and the resize issue (not bothering to try and do clever parent widget adjustSize() anymore, we never did before (e.g for the control port settings), and it causes more problems than it solves)

I'll leave to @micahflee to decide whether we should ship with a compiled version of obfs4proxy that works on all supported operating systems. I can't see a way that we can easily avoid that - maybe if we just copy the one from TorBrowser in the same way that we copy the Tor binary itself? But obviously would need an .exe for Windows, etc.

**Comment by Baccount** [comment]
@mig5 thank you for addressing the sizing issue, I'm no longer experiencing that overlap. 
Question, why does the use obfs4 pluggable transport option disable when custom bridges is enabled, but when I enable obfs4 option, the custom bridges option is enabled.

**Comment by mig5** [comment]
@Baccount  seems to be a bug in Qt, as the custom bridges radio toggle is meant to be disabled (it's in the code) but indeed it doesn't happen.

I'll see if we can do without disabling any radios altogether there instead.

**Comment by mig5** [comment]
The last commit I pushed (https://github.com/micahflee/onionshare/pull/496/commits/284213aa2d9fb4756a90891b43cdd8c974496e23) integrates the obfs4proxy binary that comes with the macOS dmg. Working perfectly on macOS

For Windows, things will be trickier: there is no separate obfs4proxy.exe shipped in the tor zip we use. We'd have to compile it ourselves or somehow extract it from the TorBrowser .exe file which I'm not sure is possible.

For Linux/BSD, I hardcode paths to obfs4proxy similar to the other hardcoded paths. I also disable the obfs4 option in the SettingsDialog if the binary does not exist in the expected location. obfs4proxy is best to be installed via Linux distribution packages same as Tor itself.

So this is more of a PoC as it doesn't add support for anything but macOS, not sure how to best address.

**Comment by Baccount** [comment]
Obfs4 works on my end, macOS 10.13.2 (17C205). Excellent job @mig5

**Comment by micahflee** [comment]
Thank you, I'm about to test this. I totally agree, for the Mac version we should just copy the binary from Tor Browser (like we do with the tor binary itself).

For the Windows version, right now the `get-tor-windows.py` script downloads the "Tor Expert Bundle", which is basically just compiled tor, which is all we really needed. And it's also easier to deal with because it's distributed as a `.zip` instead a self-extracting `.exe`. I bet I can refactor the `get-tor-windows.py` script to extract the exe though, and then grab the compiled obs4proxy binary from there. I know there's a program called cabextract that can extract compressed archives from from `.exe` files, but I'll have to do more research to see if it's available in Windows by default or if you have to install it, and if it works on the Tor Browser installer. (Worse case, I can make the Windows build script require you to pass the path of an extracted Tor Browser maybe...)

So let me review this, and try to figure out if it makes sense to merge this into the 1.2 release or save it for a later one.

**Comment by mig5** [comment]
just added obfs4proxy dependency for Linux distros, into the build/install scripts. obfs4 working for me on Fedora 26.

**Comment by mig5** [comment]
@micahflee this branch now contains the modifications to get-tor-windows.py which now fetches the TorBrowser .exe file and extracts what we need with 7-zip (BUILD.md also updated)

The end result should be that obfs4proxy.exe exists alongside the tor.exe

The only thing I'm not sure on is whether we need any other .dll files from the extracted exe

Because I can't seem to get a Tor connection going in my Windows VM, appreciate if you can do an end-to-end test of this and see if obfs4 bridges are working in your Windows installation.

**Comment by mig5** [comment]
I fixed my  Tor connection issue - when I try obfs4 on Windows I get an error '2' - must be some problem with the torrc that gets generated. Trying to work out how to debug this in a Windows environment..

**Comment by mig5** [comment]
Oh whoops, the torrc_template-obfs4 template is missing.. that'd do it. I'll fix shortly.

**Comment by mig5** [comment]
Fixed and I can see obfs4proxy running in my TaskManager.. Tor connection hanging on 20%, not sure if there are still problems under the hood or just slow.. maybe you'll be able to get this over the line..

**Comment by Baccount** [comment]
@mig5 The custom bridges field does not support Obfs4 bridges.
Here are the bridges I tested.
```
obfs4 192.36.31.144:48285 8C68078CCF04D2BD2CC110EF05336FF26020F8E5 cert=fyBsbvjklDzo/yyp22hvzb3fRWXMRlb2hLnPC5cIqr271W06L6KoGk+BVDTTqvDTlbBpUg iat-mode=0
obfs4 13.56.22.18:9443 EFF2EAC2752CE2B53F5B30D0119213675D0A5890 cert=MhcHfT0QyqEkcz9SaMqEEuLnkixzgLqp5w1MNj6FohKm4St4Hpewbft3lnxQPmrnm9rJGA iat-mode=0
obfs4 185.101.218.42:41758 E3BBFDD00B287113C7A1EB2A2C2EC1D9D6A80218 cert=ePT0ASv45frcpQ9pSzQ3Ae/wNMufga/x5jGdqdZpLGLbHdYKm3s5QQGhPkQKivRwz/Csfg iat-mode=0obfs4 192.36.31.144:48285 8C68078CCF04D2BD2CC110EF05336FF26020F8E5 cert=fyBsbvjklDzo/yyp22hvzb3fRWXMRlb2hLnPC5cIqr271W06L6KoGk+BVDTTqvDTlbBpUg iat-mode=0
obfs4 13.56.22.18:9443 EFF2EAC2752CE2B53F5B30D0119213675D0A5890 cert=MhcHfT0QyqEkcz9SaMqEEuLnkixzgLqp5w1MNj6FohKm4St4Hpewbft3lnxQPmrnm9rJGA iat-mode=0
obfs4 185.101.218.42:41758 E3BBFDD00B287113C7A1EB2A2C2EC1D9D6A80218 cert=ePT0ASv45frcpQ9pSzQ3Ae/wNMufga/x5jGdqdZpLGLbHdYKm3s5QQGhPkQKivRwz/Csfg iat-mode=0
```

**Comment by mig5** [comment]
@Baccount that's by design, custom bridges are for regular bridges. You use the obfs4 radio box for obfs4 bridges and it will use the internal obfs4 bridges that come with Tor Browser

**Comment by Baccount** [comment]
Thanks, @mig5, should it be noted at the end of the label that obfs4 can not be used? Because there's no indication that Obfs4 will not work.
`You can get bridges from https://bridges.torproject.org`

**Comment by mig5** [comment]
You can't get obfs4 bridges from https://bridges.torproject.org as far as I'm concerned - to me it seems clear given there's a separate radio box for obfs4. I leave it up to @micahflee

**Comment by Baccount** [comment]
@mig5 you can get obfs4 bridges from https://bridges.torproject.org/options
issue

**Comment by mig5** [comment]
Still can't seem to get tor to bootstrap with obfs4 via the UI on Windows. Frustratingly, via some trickery involving ```C:\Program Files (x86)\Onionshare\tor\Tor\tor.exe -f path\to\an\obf4-style\torrc | more``` via cmd.exe, I can see that it *does* bootstrap. And therefore I know that obfs4proxy.exe is actually working. But network consensus is never reached when running via the UI.. I wonder if the UI needs to 'wait' somehow for obfs4proxy to start up properly..

.. hopefully Micah can figure it out.. otherwise it still feels a bit too alpha to include in 1.2, so maybe use the other branch..

Custom bridges working just fine in Windows.. and users can always use Tor Browser hooked up to obfs4 as the Tor connection mode to achieve obfs4 connections..

File: install/get-tor-windows.py [hunk 1]
```diff
@@ -24,55 +24,53 @@
...
 import urllib.request
 
 def main():
-    zip_url = 'https://archive.torproject.org/tor-package-archive/torbrowser/7.0.11/tor-win32-0.3.1.9.zip'
-    zip_filename = 'tor-win32-0.3.1.9.zip'
-    expected_zip_sha256 = 'faf28efb606455842bda66ca369287a116b6d6e5ad3720ebed9337da0717f1b4'
-
+    exe_url = 'https://archive.torproject.org/tor-package-archive/torbrowser/7.0.11/torbrowser-install-7.0.11_en-US.exe'
+    exe_filename = 'torbrowser-install-7.0.11_en-US.exe'
+    expected_exe_sha256 = 'a033eb9b9ed2ad389169b36a90946a8af8f05bd0c7bbd3e37678041331096624'
```
Comment (micahflee) [inline install/get-tor-windows.py]: 💯

**Pull request merged**